hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/FourToFiveUpgradeHandler.java

```diff
@@ -0,0 +1,68 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.table.upgrade;
+
+import org.apache.hudi.common.config.ConfigProperty;
+import org.apache.hudi.common.config.HoodieMetadataConfig;
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieException;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.hudi.common.util.PartitionPathEncodeUtils.DEFAULT_PARTITION_PATH;
+import static org.apache.hudi.common.util.PartitionPathEncodeUtils.DEPRECATED_DEFAULT_PARTITION_PATH;
+
+/**
+ * Upgrade handler to upgrade Hudi's table version from 4 to 5.
+ */
+public class FourToFiveUpgradeHandler implements UpgradeHandler {
+
+  private static final Logger LOG = LogManager.getLogger(FourToFiveUpgradeHandler.class);
+
+  @Override
+  public Map<ConfigProperty, String> upgrade(HoodieWriteConfig config, HoodieEngineContext context, String instantTime, SupportsUpgradeDowngrade upgradeDowngradeHelper) {
+    if (FSUtils.getAllPartitionPaths(context, HoodieMetadataConfig.newBuilder().enable(false).build(), config.getBasePath())
+        .contains(DEPRECATED_DEFAULT_PARTITION_PATH)) {
+      LOG.error(String.format("\"%s\" partition detected. From 0.12, we are changing the default partition in hudi to %s "
+              + " Please read and write back the data in \"%s\" partition in hudi to new partition path \"%s\". \"\n"
+              + " Sample spark command to use to re-write the data: \n\n"
+              + " val df = spark.read.format(\"hudi\").load(HUDI_TABLE_PATH).filter(col(\"PARTITION_PATH_COLUMN\") === \"%s\"); \t \n\n"
+              + " df.drop(\"_hoodie_commit_time\").drop(\"_hoodie_commit_seqno\").drop(\"_hoodie_record_key\")\"\n"
+              + " .drop(\"_hoodie_partition_path\").drop(\"_hoodie_file_name\").withColumn(PARTITION_PATH_COLUMN,\"%s\")\"\n"
+              + " .write.options(writeOptions).mode(Append).save(HUDI_TABLE_PATH);\t\n\"\n"
+              + " Please fix values for PARTITION_PATH_COLUMN, HUDI_TABLE_PATH and set all write configs in above command before running. "
+              + " Also do delete the records in old partition once above command succeeds. "
+              + " Sample spark command to delete old partition records: \n\n"
+              + " val df = spark.read.format(\"hudi\").load(HUDI_TABLE_PATH).filter(col(\"PARTITION_PATH_COLUMN\") === \"%s\"); \t \n\n"
+              + " df.write.option(\"hoodie.datasource.write.operation\",\"delete\").options(writeOptions).mode(Append).save(HUDI_TABLE_PATH);\t\n\"\n",
```

Comment:
Do we need to add an upgrade/downgrade handler just for an option validation ? Can we just add a validation in the `HoodieWriteConfig` ?

Reply:
we don't need to validate this if not for upgrading. so don't want to execute the validation if a new table is created w/ 0.12. this is applicable only for old tables. Is there a elegant way to get this done w/o adding it as part of upgrade? let me know. would be happy to incorporate.

Reply:
1. The logic is too heavy `FSUtils.getAllPartitionPaths`, we should avoid that. 
2. And it throws exception directly, how can user fixes the exception if they want to use release 0.12.0, based on the fact that the upgrade would fail.

Can we do an auto fix when creating the new partition path ? Say when we create the new partition path, we checks the old partition if exists and if it does, rename it directly.

Reply:
just renaming the directory wound't suffice. we need to rewrite the data since we have meta fields (hoodie partition path) which needs to be fixed as well. hence we don't want to do this heavy lifting during upgrade.

Reply:
Seems no good manner for this case, can we just put the notes in release notes, and tell the user how to fix it ? To avoid to force all the users to do the upgrade in code (even though they have no partition path).

Or better we can supply a fix tool for the user.